### DISCUSSION TITLE
Backwards compatibility of keycloak-admin-client

### DISCUSSION THREAD

**Discussion by mposolda** [keycloak/keycloak#32857] [Feedback]
Until now, we have not supported backwards compatibility for Java admin-client. Using java admin-client of latest version did not work with the earlier Keycloak versions as mentioned in the [documentation](https://www.keycloak.org/docs/latest/upgrading/index.html#upgrading-the-keycloak-admin-client) . Since Keycloak 26, this is going to change and we plan to support Java admin-client even with the earlier versions of the Keycloak server. So for example it will be possible to use java admin-client 26 with the Keycloak server 24.

However there are some limitations as some admin REST API endpoints were implemented only in the newer version of Keycloak server and hence they will not work when triggered against older Keycloak server. For example Organizations feature and the related class `OrganizationResource` will work only from Keycloak 25. So if admin-client tries to invoke any related endpoints on the Keycloak 24 server, it won't work and will result in 404 error responses.

Moreover some REST endpoints require that particular feature is enabled on the server.

Starting this discussion to collect feedback on how to tell from which server version is particular method supported and which features are required. Some possible ideas:

1) Javadoc. For example something like:

```
/**
 * Supported from Keycloak 25.0.0
 * 
 * Requires {@link org.keycloak.common.Profile.Feature#ORGANIZATION} feature enabled
 */
public interface OrganizationResource {
  ...
```

2) Javadoc with custom tags. May be good for filtering etc. Something like:

```
/**
 * @sinceServerVersion 25.0.0
 * @requiresFeature {@link org.keycloak.common.Profile.Feature#ORGANIZATION}
 */
public interface OrganizationResource {
  ...
```

3) Using some custom annotation. That may not help those relying on javadoc:

```
@KeycloakServer(minVersion = "25.0.0", requiresFeatures = Profile.Feature.ORGANIZATION)
public interface OrganizationResource {
  ...
```

Any other ideas?

**Comment by yyvess** [comment]
New endpoint related to new features is not the priority for me.

1 is fine.

If we can have a client that is able to support 3 version for basic features like setup idp that will already a greate improvement.

On my point of view, compatibility should manage at servers level on rest end point to ignore unknown property, have default value for new fields and create new end-point /v2 for breaking change.
